### PR TITLE
Add unit-testing & CLI

### PR DESCRIPTION
**Pull request by defi0x1**
**PR Includes:**
- CLI: Added support for creating, updating, and closing the config.
- Unit Tests: Covered the happy path across all functions.

**TODO:**
- Implement dynamic fee updates.
- Extend CLI to support dynamic fee configuration.
- Enhance test coverage with additional edge cases.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  #     # Install rust + toolchain
+  #     - run: rustup toolchain install stable --component clippy
+  #     # Cache rust, cargo
+  #     - uses: Swatinem/rust-cache@v2
+  #       with:
+  #         prefix-key: "v1"
+  #     - run: cargo test --package cp-amm
+  #       shell: bash
+
+  cp_amm_program_test:
```
Comment (andrewsource147) [inline .github/workflows/ci.yml]: pls add CI for:
- Anchor build
- Cargo build -p cli

File: Anchor.toml [hunk 2]
```diff
@@ -9,7 +9,8 @@ url = "https://api.apr.dev"
 
 [provider]
 cluster = "Localnet"
-wallet = "/Users/andrewnguyen/.config/solana/id.json"
+# wallet = "/Users/andrewnguyen/.config/solana/id.json"
+wallet = "/Users/minhdo/.config/solana/id.json"
```
Comment (andrewsource147) [inline Anchor.toml]: pls remove personal link to wallet

File: cli/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[features]
+local = ["cp-amm/local"]
+e2e-test = ["local"]
+
+[dependencies]
+anchor-lang = { workspace = true }
+anchor-spl = { workspace = true }
+anchor-client = "0.29.0"
```
Comment (andrewsource147) [inline cli/Cargo.toml]: could you move it to workspace?

File: cli/src/instructions/close_config.rs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+use crate::common::pda::derive_event_authority_pda;
+
+#[derive(Debug)]
+pub struct CloseConfigParams {
+    pub config: Pubkey,
+    pub rent_receiver: Pubkey,
+}
+
+pub fn close_config<C: Deref<Target = impl Signer> + Clone>(
```
Comment (andrewsource147) [inline cli/src/instructions/close_config.rs:13]: pls refer logic from this function

https://github.com/MeteoraAg/meteora-dynamic-amm/blob/main/cli-amm/src/commands/create_config.rs#L29

File: cli/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                sqrt_max_price,
+                vault_config_key,
+                pool_creator_authority,
+                activation_type,
+                collect_fee_mode,
+                pool_fees: pool_fee,
+            };
+            create_config(params, &program, transaction_config, compute_unit_price_ix)?;
+        }
+        Command::UpdateConfig { config, param, value } => {
```
Comment (andrewsource147) [inline cli/src/main.rs]: pls refer how we update config
https://github.com/MeteoraAg/meteora-dynamic-amm/blob/main/cli-amm/src/commands/update_config.rs#L36

File: programs/cp-amm/src/instructions/admin/ix_create_config.rs [hunk 6]
```diff
@@ -6,19 +6,18 @@ use crate::params::customizable_params::CustomizableParams;
...
+#[derive(AnchorSerialize, AnchorDeserialize, Debug)]
 pub struct ConfigParameters {
     pub pool_fees: PoolFees,
     pub sqrt_min_price: u128,
     pub sqrt_max_price: u128,
     pub vault_config_key: Pubkey,
     pub pool_creator_authority: Pubkey,
     pub activation_type: u8,
-    pub collect_fee_mode: CollectFeeMode,
+    pub collect_fee_mode: u8,
```
Comment (andrewsource147) [inline programs/cp-amm/src/instructions/admin/ix_create_config.rs:23]: if you switch to u8, pls validate this

File: programs/cp-amm/src/instructions/admin/update_config.rs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+use crate::constants::fee::{ FEE_DENOMINATOR, MEME_MIN_FEE_NUMERATOR };
+use crate::params::pool_fees::validate_fee_fraction;
+use crate::state::config::Config;
+use crate::state::CollectFeeMode;
+use crate::PoolError;
+use anchor_lang::prelude::*;
+
+#[event_cpi]
+#[derive(Accounts)]
+pub struct UpdateConfigCtx<'info> {
```
Comment (andrewsource147) [inline programs/cp-amm/src/instructions/admin/update_config.rs]: pls remove this function

File: programs/cp-amm/src/instructions/ix_initialize_pool.rs [hunk 8]
```diff
@@ -137,7 +137,7 @@ pub struct InitializePool<'info> {
     #[account(
         mut,
         token::mint = token_a_mint,
-        token::authority = creator,
+        token::authority = payer,
         token::token_program = token_a_program,
```
Comment (andrewsource147) [inline programs/cp-amm/src/instructions/ix_initialize_pool.rs]: pls remove all macro here

File: programs/cp-amm/src/instructions/ix_initialize_pool.rs [hunk 9]
```diff
@@ -146,7 +146,7 @@ pub struct InitializePool<'info> {
     #[account(
         mut,
         token::mint = token_b_mint,
-        token::authority = creator,
+        token::authority = payer,
         token::token_program = token_b_program,
```
Comment (andrewsource147) [inline programs/cp-amm/src/instructions/ix_initialize_pool.rs]: ditto^

File: programs/cp-amm/src/lib.rs [hunk 10]
```diff
@@ -20,21 +20,34 @@ pub mod tests;
...
         instructions::handle_create_config(ctx, config_parameters)
     }
 
+    /// update config
+    pub fn update_config(ctx: Context<UpdateConfigCtx>, param: u8, value: u8) -> Result<()> {
+        instructions::handle_update_config(ctx, param, value)
+    }
+
+    /// update config
+    pub fn update_pool_fee(ctx: Context<UpdateConfigCtx>, param: u8, value: u64) -> Result<()> {
```
Comment (andrewsource147) [inline programs/cp-amm/src/lib.rs]: remove this function

File: Anchor.toml (hunk 2 above)
Comment (defi0x1) [reply Anchor.toml]: updated

File: cli/Cargo.toml (hunk 3 above)
Comment (defi0x1) [reply cli/Cargo.toml]: updated

File: cli/src/instructions/close_config.rs (hunk 4 above)
Comment (defi0x1) [reply cli/src/instructions/close_config.rs:13]: updated

File: cli/src/main.rs (hunk 5 above)
Comment (defi0x1) [reply cli/src/main.rs]: updated

File: programs/cp-amm/src/instructions/admin/ix_create_config.rs (hunk 6 above)
Comment (defi0x1) [reply programs/cp-amm/src/instructions/admin/ix_create_config.rs:23]: added

File: programs/cp-amm/src/instructions/admin/update_config.rs (hunk 7 above)
Comment (defi0x1) [reply programs/cp-amm/src/instructions/admin/update_config.rs]: removed

File: programs/cp-amm/src/instructions/ix_initialize_pool.rs (hunk 8 above)
Comment (defi0x1) [reply programs/cp-amm/src/instructions/ix_initialize_pool.rs]: removed

File: programs/cp-amm/src/instructions/ix_initialize_pool.rs (hunk 9 above)
Comment (defi0x1) [reply programs/cp-amm/src/instructions/ix_initialize_pool.rs]: removed

File: programs/cp-amm/src/lib.rs (hunk 10 above)
Comment (defi0x1) [reply programs/cp-amm/src/lib.rs]: removed

File: .github/workflows/ci.yml (hunk 1 above)
Comment (defi0x1) [reply .github/workflows/ci.yml]: updated

File: .github/workflows/ci.yml [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    runs-on: ubuntu-latest
+    needs: program_changed_files
+    if: needs.program_changed_files.outputs.program == 'true'
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ./.github/actions/setup-solana
+      - uses: ./.github/actions/setup-dep
+      - uses: ./.github/actions/setup-anchor
+
+      - run: anchor build -- --features local
```
Comment (andrewsource147) [inline .github/workflows/ci.yml]: just `anchor build`

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+      - uses: actions/checkout@v4
+      # Install rust + toolchain
+      - uses: dtolnay/rust-toolchain@stable
+        with:
+          toolchain: 1.76.0
+          components: clippy
+      - run: rustup target add x86_64-unknown-linux-gnu
+      # Cache rust, cargo
+      - uses: Swatinem/rust-cache@v1
+      - run: cargo build -p cli --features local --target=x86_64-unknown-linux-gnu
```
Comment (andrewsource147) [inline .github/workflows/ci.yml]: pls remove --features local

File: cli/src/instructions/create_config.rs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    pub collect_fee_mode: u8,
+}
+
+pub fn create_config<C: Deref<Target = impl Signer> + Clone>(
+    params: CreateConfigParams,
+    program: &Program<C>,
+    transaction_config: RpcSendTransactionConfig,
+    compute_unit_price: Option<Instruction>
+) -> Result<()> {
+    let mut index = 1u64;
```
Comment (andrewsource147) [inline cli/src/instructions/create_config.rs]: pls go from zero

File: cli/src/instructions/update_config.rs [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+) -> Result<Pubkey> {
+    let UpdateConfigParams {
+        config,
+        trade_fee_numerator,
+        protocol_fee_percent,
+        partner_fee_percent,
+        referral_fee_percent,
+    } = params;
+
+    let mut index = 0u64;
```
Comment (andrewsource147) [inline cli/src/instructions/update_config.rs]: you dont need to loop, how about to store index in config state? then you can get index from a config account?

File: programs/cp-amm/src/instructions/admin/ix_close_config.rs [hunk 15]
```diff
@@ -10,16 +10,12 @@ use anchor_lang::prelude::*;
...
-        close = rent_receiver
+        close = admin
     )]
     pub config: AccountLoader<'info, Config>,
 
     #[account(mut, constraint = assert_eq_admin(admin.key()) @ PoolError::InvalidAdmin)]
     pub admin: Signer<'info>,
-
-    /// CHECK: Account to receive closed account rental SOL
-    #[account(mut)]
```
Comment (andrewsource147) [inline programs/cp-amm/src/instructions/admin/ix_close_config.rs]: still keep rent_receiver, I mean in CLI we can set the rent_receiver same as admin address (update on CLI, not on program side)

File: programs/cp-amm/src/instructions/admin/ix_close_config.rs [hunk 16]
```diff
@@ -10,13 +10,13 @@ use anchor_lang::prelude::*;
 pub struct CloseConfigCtx<'info> {
     #[account(
         mut,
-        close = rent_receiver
+        close = admin
```
Comment (andrewsource147) [inline programs/cp-amm/src/instructions/admin/ix_close_config.rs]: close = rent_receiver
